module/data/terrain-data.mjs

```diff
@@ -0,0 +1,46 @@
+const { BooleanField } = foundry.data.fields;
+
+/**
+ * Extension of terrain data with support for 5e concepts.
+ */
+export default class TerrainData5e extends foundry.data.TerrainData {
+
+  /** @inheritDoc */
+  static defineSchema() {
+    return {
+      ...super.defineSchema(),
+      difficultTerrain: new BooleanField({ required: true })
+    };
+  }
+
+  /* -------------------------------------------- */
+
+  /** @override */
+  static resolveTerrainEffects(effects) {
+    let data = super.resolveTerrainEffects(effects);
+    if ( !effects.some(e => e.name === "difficultTerrain") ) return data;
+    if ( data ) data.updateSource({ difficultTerrain: true });
+    else data = new this({ difficultTerrain: true });
+    return data;
+  }
+
+  /* -------------------------------------------- */
+
+  /** @override */
+  static getMovementCostFunction(token, options) {
+    return (from, to, distance, segment) => {
+      let finalDistance = distance * (segment.terrain?.difficulty ?? 1);
+      if ( segment.terrain?.difficultTerrain ) finalDistance += distance;
+      return finalDistance;
+    };
+  }
```

Comment:
Alternatively, we could implement it like this because we test whether difficult terrain applies in `DifficultTerrainRegionBehaviorType#_getTerrainEffects` and therefore we can already add it to `TerrainData#difficulty`.

```suggestion
  /** @override */
  static resolveTerrainEffects(effects) {
    let data = super.resolveTerrainEffects(effects);
    if ( !effects.some(e => e.name === "difficultTerrain") ) return data;
    if ( !data ) data = new this({ difficulty: 2, difficultTerrain: true });
    else {
      let difficulty = data.difficulty + 1;
      if ( !Number.isFinite(difficulty) ) difficulty = null;
      data.updateSource({ difficulty });
    }
    return data;
  }
```

Reply:
One advantage of this is that the speed is automatically adjusted for difficult terrain in `Token#_modifyAnimationMovementSpeed` without us needing to override it.